scrapy/item.py

```diff
@@ -14,28 +14,39 @@
 from scrapy.utils.trackref import object_ref
 
 
-class BaseItem(object_ref):
-    """Base class for all scraped items.
+class _BaseItem(object_ref):
+    """
+    Temporary class used internally to avoid the deprecation
+    warning raised by isinstance checks using BaseItem.
+    """
+    pass
 
-    In Scrapy, an object is considered an *item* if it is an instance of either
-    :class:`BaseItem` or :class:`dict`. For example, when the output of a
-    spider callback is evaluated, only instances of :class:`BaseItem` or
-    :class:`dict` are passed to :ref:`item pipelines <topics-item-pipeline>`.
 
-    If you need instances of a custom class to be considered items by Scrapy,
-    you must inherit from either :class:`BaseItem` or :class:`dict`.
+class _BaseItemMeta(ABCMeta):
+    def __instancecheck__(cls, instance):
+        if cls is BaseItem:
+            warn('scrapy.item.BaseItem is deprecated, please use scrapy.item.Item instead',
+                 ScrapyDeprecationWarning, stacklevel=2)
+        return super().__instancecheck__(instance)
 
-    Unlike instances of :class:`dict`, instances of :class:`BaseItem` may be
-    :ref:`tracked <topics-leaks-trackrefs>` to debug memory leaks.
+
+class BaseItem(_BaseItem, metaclass=_BaseItemMeta):
     """
-    pass
+    Deprecated, please use :class:`scrapy.item.Item` instead
+    """
+
+    def __new__(cls, *args, **kwargs):
+        if issubclass(cls, BaseItem) and not (issubclass(cls, Item) or issubclass(cls, DictItem)):
+            warn('scrapy.item.BaseItem is deprecated, please use scrapy.item.Item instead',
+                 ScrapyDeprecationWarning, stacklevel=2)
+        return super(BaseItem, cls).__new__(cls, *args, **kwargs)
 
 
 class Field(dict):
     """Container of field metadata"""
 
 
-class ItemMeta(ABCMeta):
+class ItemMeta(_BaseItemMeta):
```

Comment:
Is this necessary?

Reply:
It is, otherwise bad things happen:
```python
In [1]: from scrapy.item import Item                                                                                                                                                                                                          
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-6456d4b4cb51> in <module>
----> 1 from scrapy.item import Item

~/scrapinghub/scrapy/scrapy/__init__.py in <module>
     32 
     33 # Declare top-level shortcuts
---> 34 from scrapy.spiders import Spider
     35 from scrapy.http import Request, FormRequest
     36 from scrapy.selector import Selector

~/scrapinghub/scrapy/scrapy/spiders/__init__.py in <module>
    111 
    112 # Top-level imports
--> 113 from scrapy.spiders.crawl import CrawlSpider, Rule  # noqa: F401
    114 from scrapy.spiders.feed import XMLFeedSpider, CSVFeedSpider  # noqa: F401
    115 from scrapy.spiders.sitemap import SitemapSpider  # noqa: F401

~/scrapinghub/scrapy/scrapy/spiders/crawl.py in <module>
     11 from scrapy.exceptions import ScrapyDeprecationWarning
     12 from scrapy.http import Request, HtmlResponse
---> 13 from scrapy.linkextractors import LinkExtractor
     14 from scrapy.spiders import Spider
     15 from scrapy.utils.python import get_func_args

~/scrapinghub/scrapy/scrapy/linkextractors/__init__.py in <module>
     14 
     15 from scrapy.utils.deprecate import ScrapyDeprecationWarning
---> 16 from scrapy.utils.misc import arg_to_iter
     17 from scrapy.utils.url import (
     18     url_is_from_any_domain, url_has_any_extension,

~/scrapinghub/scrapy/scrapy/utils/misc.py in <module>
     15 from scrapy.utils.datatypes import LocalWeakReferencedCache
     16 from scrapy.utils.python import flatten, to_unicode
---> 17 from scrapy.item import _BaseItem
     18 
     19 

~/scrapinghub/scrapy/scrapy/item.py in <module>
    136 
    137 
--> 138 class Item(DictItem, metaclass=ItemMeta):
    139     """
    140     Base class for scraped items.

TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases
```

Reply:
Metaclass magic :mage:

Reply:
:open_mouth:

TIL

Reply:
Indeed, TIL too. We don't want to be invoking Cthulhu.